pinot-core/src/main/java/com/linkedin/pinot/core/periodictask/PeriodicTaskScheduler.java

```diff
@@ -0,0 +1,126 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.periodictask;
+
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Periodic task scheduler will schedule a list of tasks based on their initial delay time and interval time.
+ */
+public class PeriodicTaskScheduler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PeriodicTaskScheduler.class);
+  private final ScheduledExecutorService _executorService;
+  private long _initialDelayInSeconds;
+  private volatile boolean _running;
+
+  private static class PeriodicTaskEntry implements Comparable<PeriodicTaskEntry> {
+    private PeriodicTask _periodicTask;
+    private long _executionTime;
+    private boolean _executed;
+
+    PeriodicTaskEntry(PeriodicTask periodicTask) {
+      _periodicTask = periodicTask;
+      _executionTime = System.currentTimeMillis() + _periodicTask.getInitialDelayInSeconds() * 1000L;
+      if (_periodicTask.getIntervalInSeconds() > 0L) {
+        _periodicTask.init();
+      }
+      _executed = false;
+    }
+
+    PeriodicTask getPeriodicTask() {
+      return _periodicTask;
+    }
+
+    long getExecutionTime() {
+      return _executionTime;
+    }
+
+    void updateExecutionTime() {
+      _executionTime = System.currentTimeMillis() + _periodicTask.getIntervalInSeconds() * 1000L;
+    }
+
+    void setExecuted(boolean isExecuted) {
+      _executed = isExecuted;
+    }
+
+    boolean getExecuted() {
+      return _executed;
+    }
+
+    @Override
+    public int compareTo(PeriodicTaskEntry o) {
+      if (this._executionTime == o._executionTime) {
+        return 0;
+      }
+      return this._executionTime < o._executionTime ? -1 : 1;
+    }
+  }
+
+  public PeriodicTaskScheduler(long initialDelayInSeconds) {
+    LOGGER.info("Initializing PeriodicTaskScheduler. Initial delay in seconds: {}", initialDelayInSeconds);
+    _executorService = Executors.newScheduledThreadPool(1);
+    _initialDelayInSeconds = initialDelayInSeconds;
+    _running = true;
+  }
+
+  public void start(List<PeriodicTask> periodicTasks) {
+    LOGGER.info("Starting PeriodicTaskScheduler.");
+    PriorityBlockingQueue<PeriodicTaskEntry> queue = new PriorityBlockingQueue<>(Math.max(1, periodicTasks.size()));
```

Comment:
If periodicTasks is empty, why even bother with the rest? We can log a message and bail out?